pallets/pallet-web3-names/src/lib.rs

```diff
@@ -0,0 +1,470 @@
+			name: Web3NameOf<T>,
+		},
+		/// A name has been banned.
+		Web3NameBanned { name: Web3NameOf<T> },
+		/// A name has been unbanned.
+		Web3NameUnbanned { name: Web3NameOf<T> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The tx submitter does not have enough funds to pay for the deposit.
+		InsufficientFunds,
+		/// The specified name has already been previously claimed.
+		Web3NameAlreadyClaimed,
+		/// The specified name does not exist.
+		Web3NameNotFound,
+		/// The specified owner already owns a name.
+		OwnerAlreadyExists,
+		/// The specified name has been banned and cannot be interacted
+		/// with.
+		Web3NameBanned,
+		/// The specified name is not currently banned.
+		Web3NameNotBanned,
+		/// The specified name has already been previously banned.
+		Web3NameAlreadyBanned,
+		/// The actor cannot performed the specified operation.
+		NotAuthorized,
+		/// A name that is too short is being claimed.
+		Web3NameTooShort,
+		/// A name that is too long is being claimed.
+		Web3NameTooLong,
+		/// A name that contains not allowed characters is being claimed.
+		InvalidWeb3NameCharacter,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Assign the specified name to the owner as specified in the
+		/// origin.
+		///
+		/// The name must not have already been claimed by someone else and the
+		/// owner must not already own another name.
+		///
+		/// Emits `Web3NameClaimed` if the operation is carried out
+		/// successfully.
+		///
+		/// # <weight>
+		/// Weight: O(1)
+		/// - Reads: Names, Owner, Banned storage entries + available currency
+		///   check + origin check
+		/// - Writes: Names, Owner storage entries + currency deposit reserve
+		/// # </weight>
+		#[pallet::weight(<T as Config>::WeightInfo::claim(name.len().saturated_into()))]
+		pub fn claim(origin: OriginFor<T>, name: Web3NameInput<T>) -> DispatchResult {
+			let origin = T::OwnerOrigin::ensure_origin(origin)?;
+			let payer = origin.sender();
+			let owner = origin.subject();
+
+			let decoded_name = Self::check_claiming_preconditions(name, &owner, &payer)?;
+
+			// No failure beyond this point
+
+			Self::register_name(decoded_name.clone(), owner.clone(), payer);
+			Self::deposit_event(Event::<T>::Web3NameClaimed {
+				owner,
+				name: decoded_name,
+			});
+
+			Ok(())
+		}
+
+		/// Release the provided name from its owner.
+		///
+		/// The origin must be the owner of the specified name.
+		///
+		/// Emits `Web3NameReleased` if the operation is carried out
+		/// successfully.
+		///
+		/// # <weight>
+		/// Weight: O(1)
+		/// - Reads: Owner storage entry + origin check
+		/// - Writes: Names, Owner storage entries + currency deposit release
+		/// # </weight>
+		#[pallet::weight(<T as Config>::WeightInfo::release_by_owner(name.len().saturated_into()))]
+		pub fn release_by_owner(origin: OriginFor<T>, name: Web3NameInput<T>) -> DispatchResult {
```

Comment:
You actually don't need to specify the name? since there is a 1to1 relationship, there is only one name to release?

Reply:
Yes, that is true for now. Good catch, I'll change that.

Reply:
https://github.com/KILTprotocol/mashnet-node/pull/314/commits/35b2651b23d9d9395a6f381d866e333b275b7f9d 👍